tests/integration/data_transformer/data_transformer_test.py

```diff
@@ -0,0 +1,379 @@
+                }
+            },
+        ],
+        [
+            DATA_TRANSFORMER_OUTPUT_FELT_CONTRACT,
+            "output_felt",
+            "outputs",
+            {"res": Exception("Not a valid type")},
+        ],
+        [
+            DATA_TRANSFORMER_OUTPUT_FELT_CONTRACT,
+            "output_felt",
+            "inputs",
+            {"arg": 3},
+        ],
+    ],
+)
+def test_from_python_fail(
+    get_abi_from_contract,
+    contract: str,
+    function_name: str,
+    mode: Literal["inputs", "outputs"],
+    data: PythonData,
+):
+    abi = get_abi_from_contract(contract)
+    with pytest.raises(DataTransformerException):
+        from_python_transformer(abi, function_name, mode)(data)
+
+
+@pytest.mark.parametrize(
+    "contract,function_name,mode,data",
+    [
+        [
+            DATA_TRANSFORMER_FELT_CONSTRUCTOR_CONTRACT,
+            "constructor",
+            "inputs",
+            [0xD],
+        ],
+        [
+            DATA_TRANSFORMER_UINT256_CONTRACT,
+            "input_uint256",
+            "inputs",
+            [214324, 2325353452],
+        ],
+        [
+            DATA_TRANSFORMER_LISTS_CONTRACT,
+            "input_lists",
+            "inputs",
+            [
+                5,  # len felt_list
+                1,
+                2,
+                3,
+                4,
+                5,
+                2,  # len uint256_list
+                134,
+                2342354235,
+                123124,
+                25245,
+            ],
+        ],
+        [
+            DATA_TRANSFORMER_TUPLE_CONTRACT,
+            "input_tuple",
+            "inputs",
+            [535345345, 61231231223],
+        ],
+        [
+            DATA_TRANSFORMER_STRUCTS_CONTRACT,
+            "input_outer_struct",
+            "inputs",
+            [13, 1, 2, 3],
+        ],
+        [
+            DATA_TRANSFORMER_OUTPUT_FELT_CONTRACT,
+            "output_felt",
+            "outputs",
+            [3732423423423],
+        ],
+        [
+            DATA_TRANSFORMER_OUTPUT_FELT_CONTRACT,
+            "output_felt",
+            "inputs",
+            [],
+        ],
+    ],
+)
+def test_to_python_pass(
+    get_abi_from_contract,
+    contract: str,
+    function_name: str,
+    mode: Literal["inputs", "outputs"],
+    data: CairoData,
+):
+    abi = get_abi_from_contract(contract)
+    to_python_transformer(abi, function_name, mode)(data)
+
+
+@pytest.mark.parametrize(
+    "contract,function_name,mode,data",
+    [
+        # AFTER RANGE CHECK
+        # [
+        #     DATA_TRANSFORMER_FELT_CONSTRUCTOR_CONTRACT,
+        #     "constructor",
+        #     "inputs",
+        #     ["very very very very long string"],
+        # ],
+        # AFTER STARKNET.py ADDS EXCEPTION ON TOO MANY VALUES
+        # [
+        #     DATA_TRANSFORMER_FELT_CONSTRUCTOR_CONTRACT,
+        #     "constructor",
+        #     "inputs",
+        #     [0x2342, 0x24],
+        # ],
```

Comment:
- Can you add links to Starknet.py issues?
- Consider using `@pytest.mark.skip` (check `gateway_facade_test.py::test_call_to_with_incorrect_args`). 
- WHY DO YOU WRITE COMMENTS IN **UPPERCASE**?

Reply:
done